### PR TITLE
xmlwf: Resolve exit status 2 ambiguity

### PR DESCRIPTION
**Pull request by hartwork**

### CODE REVIEW

File: expat/Changes [hunk 1]
```diff
@@ -5,11 +5,16 @@ NOTE: We are looking for help with a few things:
...
-                  (by adding an integer to a NULL pointer) but rather return
-                  XML_STATUS_ERROR and set the error code to (new) code
-                  XML_ERROR_NO_BUFFER. Found by UBSan (UndefinedBehaviorSanitizer)
-                  of Clang 11 (but not Clang 9).
+                    XML_GetBuffer as a user, no longer trigger undefined behavior
+                    (by adding an integer to a NULL pointer) but rather return
+                    XML_STATUS_ERROR and set the error code to (new) code
+                    XML_ERROR_NO_BUFFER. Found by UBSan (UndefinedBehaviorSanitizer)
+                    of Clang 11 (but not Clang 9).
+            #444  xmlwf: Exit status 2 was used for either:
```
Comment (timbray) [inline expat/Changes:13]: "both" is more idiomatic than "either"

File: expat/doc/xmlwf.xml [hunk 2]
```diff
@@ -423,14 +423,19 @@ supports both.
       </varlistentry>
       <varlistentry>
         <term><option>2</option></term>
-        <listitem><para>An input file was not well-formed or could not be parsed.</para>
+        <listitem><para>One or more input file were not well-formed or could not be parsed.</para>
```
Comment (timbray) [inline expat/doc/xmlwf.xml:426]: s/file/files/

**Review by timbray — COMMENTED** [review]
Um, I don't see a change to xmlwf.c to implement the new error code?

File: expat/Changes (hunk 1 above)
Comment (hartwork) [reply expat/Changes:13]: Okay, fixed with the upcoming push. Thanks!

File: expat/doc/xmlwf.xml (hunk 2 above)
Comment (hartwork) [reply expat/doc/xmlwf.xml:426]: Good catch, fixed with the upcoming push. Thanks!

**Review by hartwork — COMMENTED** [review]
> Um, I don't see a change to xmlwf.c to implement the new error code?

True, some changes didn't make it into the commit by mistake. Fixed with the upcoming push. Thanks!

**Review by timbray — APPROVED** [review]
I confess I'm a little worried about someone having xmlwf in production in a script that relies on the exit code.

**Comment by hartwork** [comment]
> I confess I'm a little worried about someone having xmlwf in production in a script that relies on the exit code.

That's why I kept the malformed input constant at 2 and moved the invalid usage case to 4. If the script fails to invoke xmlwf correctly, probably the script needs fixing anyway, no?

**Comment by hartwork** [comment]
PS: Plus exit status 2 was documented up to 2.2.10 to signal malformed input files and nothing else.

**Comment by freddrake** [comment]
Just noticed all this.

Exit status 2 is conventionally used to report invalid command-line invocations in the Unix world.

I don't recall the history of the usage specifically in **xmlwf**, which is what really matters here.

**Comment by freddrake** [comment]
That said, most users (including scripts) don't check the exact exit status, but only for success or failure.

**Comment by hartwork** [comment]
@freddrake @timbray I made the pull request with the understanding that the current situation is broken and would get better with this change rather than worse. I moved that half of exist status 2 that seemed movable without considerable damage. Are you on board with a merge of the current diff or would you like to discuss other options more, first?

**Comment by freddrake** [comment]
@hartwork: I've no objection to continuing with the current solution.  Once the exit codes became conflated, there's no way to really "win" on this.  Using the documentation as the guide to how this should be changed is as good as it's going to get.

**Comment by hartwork** [comment]
@timbray any concerns from your side with going forward and merge?

**Comment by timbray** [comment]
When we were discussing this, you did mention the case where -k is specified and then there's both an output-file-open and a not-well-formed file error, then both error codes 2 and 3 apply, but the value returned would depend on which happened last. You had mentioned the possibility of using 5 in this case. Now, this seems like a VERY unlikely scenario, but I thought I should mention it since you discussed it.

**Comment by hartwork** [comment]
@timbray we can definitely discuss if we want "first error wins", "last error wins" or some form of "collect errors" semantics with `-k` but I think that topic is (related but) orthogonal to what we have here. Let's decide about the current ambiguity of status 2 first. What do you think?

**Comment by timbray** [comment]
You're right, it's unambiguously better to sort out the 2 and 4 codes now. I'm in favor of pushing.

**Comment by hartwork** [comment]
Okay great — thanks to the two of you! :pray:

**Pull request merged**